File: TinyScript/js/src/main/js/parser/ast/Factor.ts
import ASTNode from "./ASTNode";
import Token from "../../lexer/Token";
import PeekTokenIterator from "../PeekTokenIterator";


export default class Factor extends ASTNode {

    constructor(token: Token) {
        super();
        this.lexeme = token;
        this.label = token.value;
    }

    static parse(it: PeekTokenIterator): ASTNode | null {

        const {Scalar, Variable} = require('./index')

        const token = it.peek();
        if (token?.isVariable()) {
            it.next();
            return new Variable.default(token);

        } else if (token?.isScalar()) {
            it.next();
            return new Scalar.default(token);
        }
        return null;
    }
}

File: TinyScript/java/src/main/java/paser/ast/Factor.java
package paser.ast;

import lexer.Token;
import lexer.TokenType;
import paser.util.PeekTokenIterator;

/**
 * @author MFine
 * @version 1.0
 * @date 2021/8/23 0:20
 **/
public class Factor extends ASTNode {
    protected Factor(ASTNode parent, PeekTokenIterator it) {
        super();
        Token next = it.next();
        TokenType type = next.getType();
        if (type == TokenType.VARIABLE) {
            this.type = ASTNodeTypes.VARIABLE;
        } else {
            this.type = ASTNodeTypes.SCALAR;
        }
        this.label = next.getValue();
        this.lexeme = next;

    }

    public static ASTNode parse(ASTNode parent, PeekTokenIterator it) {

        Token next = it.peek();
        if (next.isVariable()) {
            return new Variable(parent, it);
        } else if (next.isScalar()) {
            return new Scalar(parent, it);
        }
        return null;
    }
}
